data_layer: Simplify wallet get and create code

xdustinface opened the pull request:
<!-- Merging Requirements:
- Please give your PR a title that is release-note friendly
- In order to be merged, you must add the most appropriate category Label (Added, Changed, Fixed) to your PR
-->
<!-- Explain why this is an improvement (Does this add missing functionality, improve performance, or reduce complexity?) -->
### Purpose:

Just some cleanup, see individual commits.

### New Behavior:

No change in behaviour.

xdustinface committed:
Simplify `DataLayerWallet.create_new_dl_wallet`

xdustinface committed:
Apply `final` decorator to `DataLayerWallet`

xdustinface committed:
Simplify `DataLayerWallet.create`

arvidn commented on chia/data_layer/data_layer_wallet.py:
```diff
@@ -138,34 +135,26 @@ def id(self) -> uint32:
+                ), f"WalletType.DATA_LAYER should be a DataLayerWallet instance got: {type(wallet).__name__}"
+                return wallet
+
+        async with wallet_state_manager.lock:
```
the lock is new, right? and looks like it's fixing possible races. If any `async` call is introduced above though, we'd have a problem. It looks like it would be safer to acquire the lock before looking for an existing wallet.

e.g. if any call is made to yield while looking for the existing wallet, another task can create it without us noticing, and we then create it again

arvidn reviewed (dismissed):
looks good. I think it would make it a bit more robust if you acquire the mutex before looking for an existing wallet though

xdustinface commented:
@arvidn Changed it again to lock only if its not already locked and release only if it wasn't locked because it is in some cases also called within existing lock contexts.

xdustinface committed:
Improve `WalletStateManager.get_dl_wallet`

xdustinface committed:
More use of `WalletStateManager.get_dl_wallet`

xdustinface commented:
@arvidn Okay so i dropped the idea of merging the get/create code into `DataLayerWallet.get_or_create` because we not necessarily need to lock the wallet state manager in most cases where we are only interested in finding the wallet in `WalletStateManager.wallets` imo. And merging the code would add those cases under the lock which would change behaviour of data layer RPC endpoints and one case during state processing. So, instead i now reverted those commits and added [cd895df](https://github.com/Chia-Network/chia-blockchain/pull/14878/commits/cd895df700368c163665594f30d905cf9d671e4a) and  [23ec318](https://github.com/Chia-Network/chia-blockchain/pull/14878/commits/23ec318c760c1a4e32df53df33043d2a4e8a092e).

altendky commented on chia/rpc/wallet_rpc_api.py:
```diff
@@ -3120,18 +3120,11 @@ async def create_new_dl(self, request) -> Dict:
-                dl_wallet = wallet
-                break
-        else:
+        try:
+            dl_wallet = self.service.wallet_state_manager.get_dl_wallet()
+        except ValueError:
```
I think this could be in the lock?  Or am I misunderstanding the existing comments about locking etc.  At which point basically just the whole function could be in a single acquisition of the lock.

xdustinface replied:
Kind of yes, for the very first time its called there is a race and but after i guess it doesn't matter because there always will be only DL wallet. But still it seems there might be some cases where we could get "size changed during iteration" issues... However, this PR just retains behaviour while cleaning up some duplicated code now, moving it into the lock means those endpoints will block also for wallet fetching so i think instead of just locking there, we should cache the DL wallet separately in the wallet state manager since it can only be one and just access the cached wallet.

altendky commented on chia/wallet/wallet_state_manager.py:
```diff
@@ -1745,9 +1739,11 @@ async def convert_puzzle_hash(self, wallet_id: uint32, puzzle_hash: bytes32) ->
-                assert isinstance(wallet, DataLayerWallet)
+                assert isinstance(
+                    wallet, DataLayerWallet
+                ), f"WalletType.DATA_LAYER should be a DataLayerWallet instance got: {type(wallet).__name__}"
```
thanks!

altendky commented on chia/wallet/wallet_state_manager.py:
```diff
@@ -1745,9 +1739,11 @@ async def convert_puzzle_hash(self, wallet_id: uint32, puzzle_hash: bytes32) ->
+                ), f"WalletType.DATA_LAYER should be a DataLayerWallet instance got: {type(wallet).__name__}"
                 return wallet
-        return None
+        raise ValueError("DataLayerWallet not available")
```
Could afford to be it's own error, but sure, this fits with existing patterns.

xdustinface replied:
Yeah i just moved it into there from all the outer places, we use the `ValueError` in a bunch of other related places so thats a separate PR.

altendky commented on chia/wallet/wallet_state_manager.py:
```diff
@@ -1303,15 +1301,11 @@ async def add_coin_states(
+                                    try:
+                                        dl_wallet = self.get_dl_wallet()
+                                    except ValueError:
                                         dl_wallet = await DataLayerWallet.create_new_dl_wallet(
```
Isn't this a call that should be made under a wallet state manager lock but is not?

altendky commented on chia/wallet/wallet_state_manager.py:
```diff
@@ -1047,14 +1043,16 @@ async def add_coin_states(
                             if (
-                                await potential_dl.get_singleton_record(coin_name) is not None
+                                await dl_wallet.get_singleton_record(coin_name) is not None
                                 or coin_state.coin.puzzle_hash == MIRROR_PUZZLE_HASH
                             ):
-                                wallet_id = potential_dl.id()
-                                wallet_type = potential_dl.type()
+                                wallet_id = dl_wallet.id()
+                                wallet_type = dl_wallet.type()
```
This probably belongs in the `else:` or such so you don't start consuming `ValueError` from it.  `try:` blocks should be minimal.

xdustinface replied:
See 471aedc9778bff68ee48e11ac23fb000742d4e95.

altendky reviewed (changes requested):
The locking comments can be ignored for this PR.  We should handle them at some point though so I at least made notes.  There are a couple other points to be addressed here.

xdustinface committed:
Only try `get_dl_wallet`

xdustinface commented on chia/rpc/wallet_rpc_api.py:
```diff
@@ -3191,119 +3176,93 @@ async def dl_latest_singleton(self, request) -> Dict:
-        raise ValueError("No DataLayer wallet has been initialized")
+        only_confirmed = request.get("only_confirmed")
+        if only_confirmed is None:
+            only_confirmed = False
```
Hm, how is `only_confirmed` related here?

altendky replied:
A couple points.  `request` is unhinted and thus either this method is either not analyzed or `request` is `Any`.  I think the latter since there is a `Dict` return type hint.  Even if `request` were hinted, we hint as `Dict[str, Any]` so `only_confirmed` would still be `Any`.

This is the mess that will remain even if we ever get through the existing mypy loose-checking module list.
